lib/streamlit/runtime/scriptrunner/script_runner.py

```diff
@@ -446,24 +445,33 @@ def _run_script(self, rerun_data: RerunData) -> None:
 
             fragment_ids_this_run = set(rerun_data.fragment_id_queue)
 
+            ctx = self._get_script_run_ctx()
             # Clear widget state on page change. This normally happens implicitly
             # in the script run cleanup steps, but doing it explicitly ensures
             # it happens even if a script run was interrupted.
             try:
-                old_hash = self._session_state[SCRIPT_RUN_PAGE_SCRIPT_HASH_KEY]
+                old_hash = ctx.page_script_hash
             except KeyError:
                 old_hash = None
 
             if old_hash != page_script_hash:
-                # Page changed, reset widget state
-                self._session_state.on_script_finished(set())
+                # Page changed, reset widget state where possible. Use the
+                # widget ids from the rerun data to maintain some widget state
```

Comment:
I think it would be beneficial to explain briefly _why_ and the scenario where this can happen. If someone will look at this in a year from now they might wonder: why do we retain widget ids and which one?
The comment should give a brief explanation to this question.

```python
# Page changed, reset widget state where possible. Use the
# widget ids from the rerun data to maintain some widget state.
# When a user navigates from page A to page B, we keep the widget_ids in the rerun_data because ...
```